src/PowerShellEditorServices/Workspace/WorkspaceFileSystemWrapper.cs

```diff
@@ -0,0 +1,317 @@
+            }
+        }
+    }
+
+    /// <summary>
+    /// Wraps an instance of <see cref="System.IO.DirectoryInfo" /> and provides implementation of
+    /// <see cref="Microsoft.Extensions.FileSystemGlobbing.Abstractions.DirectoryInfoBase" />.
+    /// Based on https://github.com/aspnet/Extensions/blob/c087cadf1dfdbd2b8785ef764e5ef58a1a7e5ed0/src/FileSystemGlobbing/src/Abstractions/DirectoryInfoWrapper.cs
+    /// </summary>
+    public class WorkspaceFileSystemDirectoryWrapper : DirectoryInfoBase
+    {
+        private readonly DirectoryInfo _concreteDirectoryInfo;
+        private readonly bool _isParentPath;
+        private readonly WorkspaceFileSystemWrapperFactory _factory;
+        private readonly int _depth;
+
+        /// <summary>
+        /// Initializes an instance of <see cref="WorkspaceFileSystemDirectoryWrapper" />.
+        /// </summary>
+        public WorkspaceFileSystemDirectoryWrapper(WorkspaceFileSystemWrapperFactory factory, DirectoryInfo directoryInfo, int depth)
+        {
+            _concreteDirectoryInfo = directoryInfo;
+            _isParentPath = (depth == 0);
+            _factory = factory;
+            _depth = depth;
+        }
+
+        /// <inheritdoc />
+        public override IEnumerable<FileSystemInfoBase> EnumerateFileSystemInfos()
+        {
+            if (!_concreteDirectoryInfo.Exists || _depth >= _factory.MaxRecursionDepth) { yield break; }
+            foreach (FileSystemInfo fileSystemInfo in _factory.SafeEnumerateFileSystemInfos(_concreteDirectoryInfo))
+            {
+                switch (fileSystemInfo)
+                {
+                    case DirectoryInfo dirInfo:
+                        yield return _factory.CreateDirectoryInfoWrapper(dirInfo, _depth + 1);
+                        break;
+                    case FileInfo fileInfo:
+                        yield return _factory.CreateFileInfoWrapper(fileInfo, _depth);
+                        break;
+                    default:
+                        // We should NEVER get here, but if we do just continue on
+                        break;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns an instance of <see cref="DirectoryInfoBase" /> that represents a subdirectory.
+        /// </summary>
+        /// <remarks>
+        /// If <paramref name="name" /> equals '..', this returns the parent directory.
+        /// </remarks>
+        /// <param name="name">The directory name.</param>
+        /// <returns>The directory</returns>
+        public override DirectoryInfoBase GetDirectory(string name)
+        {
+            var isParentPath = string.Equals(name, "..", StringComparison.Ordinal);
+
+            if (isParentPath) { return ParentDirectory; }
+
+            var dirs = _concreteDirectoryInfo.GetDirectories(name);
+
+            if (dirs.Length == 1) { return _factory.CreateDirectoryInfoWrapper(dirs[0], _depth + 1); }
+            if (dirs.Length == 0) { return null; }
```

Comment:
I'd put this first, since it's a simpler case

Reply:
Yeah, but, given the nature of the method, you're more likely to hit the first if clause (looking for a directory) rather than it doesn't exist (the second if clause)